File: superpermutation.rb
class Superpermutation
  def self.calc(val)
    a=(1..val.to_i).to_a
    products = a.permutation.to_a.map {|x| x.join('')} # ["12", "21"]
    permutation_of_products = products.permutation.to_a # [["12","21"], ["21", "12"]]
    shortest_string = nil

    permutation_of_products.each do |products| # ["12", "21"]
      test_string = products.reduce('') do |sum, product| # "12"
        if sum[/#{product}/]
          # string already contains combination - do nothing
          sum
        else
          # concat new pattern
          sum.concat(product)

          # Remove adjacent dups
          sum_array = sum.split('') # ["1","2","2","1"]
          unique_array = []
          previous_val = nil
          sum_array.each do |x|
            if (previous_val != x)
              unique_array.push(x)
            end
            previous_val = x
          end
          sum = unique_array.join('') # "121"
        end
        sum
      end

      if shortest_string.nil? || (shortest_string.length > test_string.length)
        shortest_string = test_string
      end
    end

    shortest_string
  end

  def self.pretty(val)
    puts "Shortest string is '#{calc(val)}'"
  end
end

File: superpermutation.js
module.exports = (() => {
  function calc(val) {

    const cluster = require('cluster');

    if (cluster.isMaster) {
      let i = 1;
      const a = Array.from({length: val}, () => i++);
      let products = [];
      let generator1 = permutations(a);
      while((p = generator1.next().value) !== undefined) {
        products.push(p);
      }
      products = products.map((x) => x.join('')); // ["12", "21"]

      let generator2 = permutations(products); // [['12', '21'], ['21', '12']]
      let shortestString = null;

      // Worker setup
      const numCPUs = require('os').cpus().length;

      // Fork workers.
      for (let i = 0; i < numCPUs; i++) {
        let worker = cluster.fork();

        // Receive message from workers
        worker.on('message', (testString) => {
          // console.log(`From worker ${testString}`);
          if (shortestString == null || (shortestString.length > testString.length)) {
            shortestString = testString;
            console.log(`New shortest superpermutation: '${shortestString}'`);
          }
        });
      }
      // End worker setup

      // Send message to workers
      while ((p = generator2.next().value) != undefined) { // ['12', '21']
        let id = Math.floor(Math.random() * numCPUs) + 1;
        cluster.workers[id].send(p);
      };

      for (const id in cluster.workers) {
        cluster.workers[id].kill();
      }
      // return `Final result ${shortestString}`;

    } else {
      process.on('message', (p) => {
        let data = buildTestString(p);
        process.send(data);
      });
      // cluster.worker.kill();
      return 'Worker exiting';
    }
  }

  function pretty(val) {
    return `Shortest string is '${calc(val)}'`;
  }

  // https://lowrey.me/permutation-with-an-es6-javascript-generator-2/
  function* permutations(elements) {
    if (elements.length === 1) {
      yield elements;
    } else {
      let [first, ...rest] = elements;
      for (let perm of permutations(rest)) {
        for (let i = 0; i < elements.length; i++) {
          let start = perm.slice(0, i);
          let rest = perm.slice(i);
          yield [...start, first, ...rest];
        }
      }
    }
  };

  function buildTestString(p) {
    return(p.reduce((sum, product) => { // '12'
      if (sum.search(product) > -1) {
        // string already contains combination - do nothing
        return sum;
      } else {
        // concat new pattern
        sum = sum.concat(product);

        // remove adjacent dups
        let sumArray = sum.split(''); // ['1','2','2','1']
        let uniqueArray = [];
        let previousVal = null;

        sumArray.forEach((x) => {
          if (previousVal !== x) {
            uniqueArray.push(x);
          }
          previousVal = x;
        });
        sum = uniqueArray.join(''); // '121'
      }
      return sum;
    }, ''));
  };

  return {
    calc: calc,
    pretty: pretty,
  }
})();